akka-projection-jdbc/src/main/scala/akka/projection/jdbc/internal/JdbcOffsetStore.scala

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.projection.jdbc.internal
+
+import java.sql.Connection
+import java.sql.Timestamp
+import java.time.Clock
+import java.time.Instant
+
+import scala.concurrent.Future
+
+import akka.Done
+import akka.annotation.InternalApi
+import akka.projection.ProjectionId
+import akka.projection.internal.OffsetSerialization.SingleOffset
+import akka.projection.internal.OffsetSerialization.fromStorageRepresentation
+import akka.projection.internal.OffsetSerialization.toStorageRepresentation
+import akka.projection.jdbc.internal.DialectDefaults.InsertIndices
+import akka.projection.jdbc.javadsl.JdbcSession
+import akka.projection.jdbc.javadsl.JdbcSession.tryWithResource
+import akka.projection.jdbc.javadsl.JdbcSession.withConnection
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] class JdbcOffsetStore[S <: JdbcSession](
+    settings: JdbcSettings,
+    jdbcSessionFactory: () => S,
+    clock: Clock) {
+
+  def this(settings: JdbcSettings, jdbcSessionFactory: () => S) =
+    this(settings, jdbcSessionFactory, Clock.systemUTC())
+
+  private[akka] implicit val executionContext = settings.executionContext
+
+  def createIfNotExists(): Future[Done] = {
+    withConnection(jdbcSessionFactory) { conn =>
+      tryWithResource(conn.createStatement()) { stmt =>
+        stmt.execute(settings.dialect.createTableStatement)
+        stmt.execute(settings.dialect.alterTableStatement)
+        Done
+      }
+    }
+  }
+
+  def clearOffset(projectionId: ProjectionId): Future[Done] =
+    withConnection(jdbcSessionFactory) { conn =>
+      tryWithResource(conn.prepareStatement(settings.dialect.clearOffsetStatement)) { stmt =>
+        stmt.setString(1, projectionId.name)
+        stmt.setString(2, projectionId.key)
+        stmt.executeUpdate()
+        Done
+      }
+    }
+
+  def readOffset[Offset](projectionId: ProjectionId): Future[Option[Offset]] =
+    withConnection(jdbcSessionFactory) { conn =>
+      tryWithResource(conn.prepareStatement(settings.dialect.readOffsetQuery)) { stmt =>
+        stmt.setString(1, projectionId.name)
+
+        val resultSet = stmt.executeQuery()
+
+        // FIXME: need to iterate over all results to build MultipleOffsets
```

Comment:
might want to fix that since it will give surprising results otherwise

Reply:
I'm deferring that to when we support MultipleOffsets

Reply:
surre, but be aware of that it can give very wrong results for ordinary offsets now if using more than one projectionId

Reply:
The query is by projectionId, so at the moment you can only return one or zero results.

Reply:
it's not! it on the `projectionId.name`

Reply:
In that case, that's not correct. 

Since we are not yet supporting MultipleOffsets, we should filter by key as well. I will move in that direction because it the least effort for now. 

MultipleOffsets support will come in following PRs. I don't want to squeeze everything here.